HTTP Sec-WebSocket-* headers (#35606)
mdn/content: files/en-us/web/http/headers/sec-websocket-extensions/index.md

---
title: Sec-WebSocket-Extensions
slug: Web/HTTP/Headers/Sec-WebSocket-Extensions
page-type: http-header
browser-compat: http.headers.Sec-WebSocket-Extensions
spec-urls: https://datatracker.ietf.org/doc/html/rfc6455#section-11.3.2
---

{{HTTPSidebar}}

> The **Sec-WebSocket-Extensions** HTTP {{glossary("request header", "request")}} and {{glossary("response header")}} is used in the [WebSocket](/en-US/docs/Web/API/WebSockets_API) opening [handshake](/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake) to negotiate a protocol extension used by the client and server.

Author: Note, not really sure what this is for. I mean I get the purpose, but I can't see how it is set via the Web API.

Reviewer: If you do `const socket = new WebSocket("ws://127.0.0.1:8081");` in the console in Chrome, you get 

```http
GET ws://127.0.0.1:8081/ HTTP/1.1
Host: 127.0.0.1:8081
Connection: Upgrade
Pragma: no-cache
Cache-Control: no-cache
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36
Upgrade: websocket
Origin: http://127.0.0.1:8080
Sec-WebSocket-Version: 13
Accept-Encoding: gzip, deflate, br, zstd
Accept-Language: en-US,en;q=0.9
Sec-WebSocket-Key: DGPhzD/pJrkMuL4p0VOkYw==
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
```

So I don't think you set this via API at all, it's hidden away and taken care for you by the underlying implementation.

Author: Thank you. I think it is worth stating, given that I asked. Done in https://github.com/mdn/content/pull/35606/files#r1746307155

In a request the header specifies one or more extensions that the web application would like to use, in order of preference.
These can be added as in multiple headers, or as comma separate values added to a single header.

> In a response the header can only appear once, where it specifies the extension selected by the server.

Reviewer: ```suggestion
In a response the header can only appear once, where it specifies the extension selected by the server from the client's preferences.
```

> This value must be the first extension that the server supports from the list provided in the request header.
> 
> <table class="properties">

Author: ```suggestion
This value must be the first extension that the server supports from the list provided in the request header.

The request header is automatically added by the browser based on the sub-protocols that it supports.

<table class="properties">
```

  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>{{Glossary("Request header")}}, {{Glossary("Response header")}}</td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden header name")}}</th>
      <td>yes</td>
    </tr>
  </tbody>
</table>

## Syntax

Request

```http
Sec-WebSocket-Extensions: <extensions>
```

Response

```http
Sec-WebSocket-Extensions: <selected-extension>
```

## Directives

- `<extensions>`
  - : A comma-separated list of extensions to request (or for the server to agree to support).
    These should be selected from the [IANA WebSocket Extension Name Registry](https://www.iana.org/assignments/websocket/websocket.xml#extension-name).
    Extensions which take parameters delineate them with semicolons.

## Examples

> The HTTP request below shows the opening handshake where a client supports the superspeed and colormode extensions.

Reviewer: ```suggestion
The HTTP request below shows the opening handshake where a client supports the `permessage-deflate` and `client_max_window_bits` extensions.
```

Author: Yeah, much better given that these are used in the spec.

```http
GET /chat HTTP/1.1
Host: example.com:8000
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==
Sec-WebSocket-Version: 13

> Sec-WebSocket-Extensions: superspeed, colormode; depth=16

Reviewer: Do you want to use `permessage-deflate` and `client_max_window_bits`?

```suggestion
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
```

```

The request below with separate headers for each extension is equivalent:

```http
GET /chat HTTP/1.1
Host: example.com:8000
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==
Sec-WebSocket-Version: 13

> Sec-WebSocket-Extensions: superspeed
> Sec-WebSocket-Extensions: colormode; depth=16

Reviewer: ```suggestion
Sec-WebSocket-Extensions: permessage-deflate
Sec-WebSocket-Extensions: client_max_window_bits
```

```

> The response below might be sent from a server to indicate that it will support the `superspeed` extension:

Reviewer: ```suggestion
The response below might be sent from a server to indicate that it will support the `permessage-deflate` extension:
```

```http
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: s3pPLMBiTxaQ9kYGzzhZRbK+xOo=

> Sec-WebSocket-Extensions: superspeed

Reviewer: ```suggestion
Sec-WebSocket-Extensions: permessage-deflate
```
